orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py

```diff
@@ -356,17 +355,120 @@ def test_input_requires_grad_backward_creates_input_grad(device):
     s.backward()
     assert x.grad is not None
 
-# TODO: Re-enable this Test when .to(), .cpu() and .cuda() are fixed
-# @pytest.mark.parametrize("device", ['cuda', 'cpu'])
-# def test_changes_input_requires_grad_reinitializes_module_gradient_graph_builder(device):
-#     N, D_in, H, D_out = 32, 784, 500, 10
-#     model = NeuralNetSinglePositionalArgument(D_in, H, D_out).to(device)
-#     model = ORTModule(model)
-#     x = torch.randn(N, D_in, device=device, requires_grad=True)
-#     model(x.data)
-#     module_gradient_graph_builder = model._module_gradient_graph_builder
-#     model(x)
-#     assert module_gradient_graph_builder != model._module_gradient_graph_builder
+def test_multiple_forward_only_calls():
+    N, D_in, H, D_out = 32, 784, 500, 10
+    model = NeuralNetSinglePositionalArgument(D_in, H, D_out).to('cuda')
+    model = ORTModule(model)
+    for step in range(10):
+        x = torch.randn(N, D_in, device='cuda', requires_grad=False)
+        prediction1 = model(x)
+
+def test_multiple_overlapping_forward_backward_calls():
+    N, D_in, H, D_out = 32, 784, 500, 10
+    model = NeuralNetSinglePositionalArgument(D_in, H, D_out).to('cuda')
+    model = ORTModule(model)
+
+    for step in range(10):
+        x1 = torch.randn(N, D_in, device='cuda', requires_grad=True)
+        x2 = torch.randn(N, D_in, device='cuda', requires_grad=True)
+        assert x1.grad is None and x2.grad is None
+        
+        prediction1 = model(x1)
+        s1 = prediction1.sum()
+
+        prediction2 = model(x2)
+        s2 = prediction2.sum()
+
+        s1.backward()
+        s2.backward()
+        assert x1.grad is not None and x2.grad is not None
+
+def test_multiple_ortmodules_training():
```

Comment:
>test_multiple_ortmodules_training [](start = 4, length = 33)

Can we assume this is equivalent to Mainz's multi-tasking case, or should have another test case?   
fw 1
bw 1
fw 2
bw 2

Reply:
yes. done.